graphql_compiler/macros/macro_expansion.py

```diff
@@ -28,21 +29,144 @@ def _merge_non_overlapping_dicts(merge_target, new_data):
+                    raise AssertionError(u'Cannot prove type coercion at macro target is valid. '
+                                         u'Please provide a proof that {} subclasses {} using the '
+                                         u'subclass_sets argument.'
+                                         .format(coercion_class, target_class))
+                else:
+                    if (target_class not in subclass_sets or
+                            coercion_class not in subclass_sets[target_class]):
+                        raise GraphQLCompilationError(u'Invalid type coercion at macro target. {} '
+                                                      u'is expected to subclass {}.'
+                                                      .format(coercion_class, target_class))
+        else:
+            # TODO(bojanserafimov): When compiling the macro, compute the type at the
+            #                       target and record that in the macro descriptor.
+            raise NotImplementedError(u'Cannot coerce from macro that does not end '
+                                      u'in a coercion.')
+
+    # Merge the continuation into the target
+    target_ast.directives += continuation_ast.directives
+    target_ast.selection_set = _merge_selection_sets(
+        target_ast.selection_set, continuation_ast.selection_set)
+    target_ast.directives = [
+        directive
+        for directive in target_ast.directives
+        if directive.name.value != MacroEdgeTargetDirective.name
+    ]
+
+
+def _expand_specific_macro_edge(macro_selection_set, selection_ast, subclass_sets=None):
     """Produce a tuple containing the new replacement selection AST, and a list of extra selections.
 
     Args:
-        macro_definition_selection_set: SelectionSet GraphQL object containing the selections
-                                        defining the macro edge. Can be retrieved as the
-                                        "expansion_selection_set" key from a MacroEdgeDescriptor.
+        macro_selection_set: SelectionSet GraphQL object containing the selections defining the
+                             macro edge. Can be retrieved as the "expansion_selection_set" key from
+                             a MacroEdgeDescriptor.
         selection_ast: GraphQL AST object containing the selection that is relying on a macro edge.
         subclass_sets: optional dict mapping class names to the set of its subclass names
 
     Returns:
         (replacement selection AST, list of extra selections that need to be added and merged)
         The first value of the tuple is a replacement
     """
-    raise NotImplementedError()
+    replacement_selection_ast = None
+    extra_selections = []
+
+    # TODO(bojanserafimov): Rename macro tags if conflicting with user-defined tag names.
+    # TODO(bojanserafimov): Remove macro tags if the user has tagged the same field.
+
+    for macro_ast in deepcopy(macro_selection_set).selections:
```

Comment:
Let's avoid this `deepcopy` if at all possible. Copying ASTs can get really expensive since they are highly nested and contain a very large number of objects.

Reply:
Removing this one is not simple at all. And even if we removed it, we would need to shallow copy all objects on the path to the `@macro_edge_target`. I'm not convinced that's any faster than a `deepcopy`. Also if we did that, changing the result of `expand_macros_in_query_ast` would change the macro definition in the registry, which is asking for bugs.